src/persist-client/src/internal/encoding.rs

```diff
@@ -62,6 +67,103 @@ impl<K: Codec, V: Codec> Clone for Schemas<K, V> {
     }
 }
 
+/// A proto that is decoded on use.
+///
+/// Because of the way prost works, decoding a large protobuf may result in a
+/// number of very short lived allocations in our RustType/ProtoType decode path
+/// (e.g. this happens for a repeated embedded message). Not every use of
+/// persist State needs every transitive bit of it to be decoded, so we opt
+/// certain parts of it (initially stats) to be decoded on use.
+///
+/// This has the dual benefit of only paying for the short-lived allocs when
+/// necessary and also allowing decoding to be gated by a feature flag. The
+/// tradeoffs are that we might decode more than once and that we have to handle
+/// invalid proto errors in more places.
+///
+/// Mechanically, this is accomplished by making the field a proto `bytes` types
+/// instead of `ProtoFoo`. These bytes then contain the serialization of
+/// ProtoFoo. NB: Swapping between the two is actually a forward and backward
+/// compatible change.
+///
+/// > Embedded messages are compatible with bytes if the bytes contain an
+/// > encoded version of the message.
+///
+/// (See <https://protobuf.dev/programming-guides/proto3/#updating>)
+#[derive(Clone, Serialize, Deserialize)]
+pub struct LazyProto<T> {
+    buf: Bytes,
+    _phantom: PhantomData<fn() -> T>,
+}
+
+impl<T: Message + Default> Debug for LazyProto<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self.decode() {
+            Ok(proto) => Debug::fmt(&proto, f),
+            Err(err) => f
+                .debug_struct(&format!("LazyProto<{}>", std::any::type_name::<T>()))
+                .field("err", &err)
+                .finish(),
+        }
+    }
+}
+
+impl<T> PartialEq for LazyProto<T> {
+    fn eq(&self, other: &Self) -> bool {
```

Comment:
(This changes equality semantic slightly for structs that contain unordered fields like maps, I think, but that seems fine to me.)